web/src/topic_list_data.js

```diff
@@ -3,37 +3,26 @@ import * as resolved_topic from "../shared/src/resolved_topic";
 import * as hash_util from "./hash_util";
 import * as narrow_state from "./narrow_state";
 import * as stream_topic_history from "./stream_topic_history";
+import * as sub_store from "./sub_store";
 import * as unread from "./unread";
 import * as user_topics from "./user_topics";
 import * as util from "./util";
 
 const max_topics = 8;
 const max_topics_with_unread = 12;
 
-export function get_list_info(stream_id, zoomed, search_term) {
-    let topics_selected = 0;
-    let more_topics_unreads = 0;
-    let more_topics_have_unread_mention_messages = false;
-
+function choose_topics(topic_names, zoomed, items, stream_info) {
     let active_topic = narrow_state.topic();
-
     if (active_topic) {
         active_topic = active_topic.toLowerCase();
     }
-
-    let topic_names = stream_topic_history.get_recent_topic_names(stream_id);
-    if (zoomed) {
-        topic_names = util.filter_by_word_prefix_match(topic_names, search_term, (item) => item);
-    }
-
-    const items = [];
-
-    const topics_with_unread_mentions = unread.get_topics_with_unread_mentions(stream_id);
-
+    const topics_with_unread_mentions = unread.get_topics_with_unread_mentions(
+        stream_info.stream_id,
+    );
```

Comment:
Let's pass this value in from the caller -- there's no reason to compute it twice, which is what your algorithm here does.